packages/core/src/hooks/useRawLogs.ts

```diff
@@ -25,21 +26,82 @@ export function useRawLogs(
   const blockNumbers = useBlockNumbers()
 
   const [logs, setLogs] = useState<Log[] | undefined>()
+  const [lastContractAddress, setLastContractAddress] = useState<string | undefined>()
+  const [lastTopics, setLastTopics] = useState<string | undefined>()
+  const [lastChainId, setLastChainId] = useState<ChainId | undefined>()
+  const [lastBlockNumber, setLastBlockNumber] = useState<number | undefined>()
+  const [isLoading, setIsLoading] = useState<boolean>(false)
 
-  const { chainId } = queryParams
+  const { chainId, isStatic } = queryParams
+  const config = useConfig()
+  const refresh = queryParams?.refresh ?? config.refresh
 
   const [provider, blockNumber] = useMemo(
     () => (chainId ? [providers[chainId], blockNumbers[chainId]] : [library, _blockNumber]),
     [providers, library, blockNumbers, _blockNumber, chainId]
   )
 
+  const deps: any[] = [provider]
+
+  const filterTopicsAsJson = filter && JSON.stringify(filter.topics)
+
+  // Push the filter elements to the dependencies. We do this individually b/c hook dependency checks are shallow
+  deps.push(filter && filter.address)
+  deps.push(filterTopicsAsJson)
+  deps.push(filter && (filter as FilterByBlockHash).blockHash)
+  deps.push(filter && (filter as Filter).fromBlock)
+  deps.push(filter && (filter as Filter).toBlock)
+
+  // Push the block number if we are not static
+  deps.push(!isStatic && refresh !== 'never' ? blockNumber : 0)
+
   async function updateLogs() {
-    setLogs(!filter ? undefined : await provider?.getLogs(filter))
+    if (isLoading) {
+      // We are already loading, don't start another request
+      return
+    }
+
+    setIsLoading(true)
+    try {
+      let filterChanged = true
+      if (
+        chainId === lastChainId &&
+        filter &&
+        lastContractAddress === filter.address &&
+        lastTopics === filterTopicsAsJson
+      ) {
+        // The filter did not change
+        filterChanged = false
+      } else {
+        // Filter changed. Reset logs
+        setLogs(undefined)
+      }
+
+      if (!filterChanged) {
+        if (isStatic || refresh === 'never') {
+          // Only update logs if contract address or topics changed
+          return
+        } else if (typeof refresh === 'number') {
+          // Only update logs if the block number has increased by the refresh interval
+          if (blockNumber && lastBlockNumber && blockNumber - lastBlockNumber < refresh) {
+            return
+          }
+        }
+      }
+
+      setLogs(!filter ? undefined : await provider?.getLogs(filter))
+      setLastContractAddress(!filter ? undefined : filter.address)
+      setLastTopics(!filter ? undefined : JSON.stringify(filter.topics))
+      setLastChainId(chainId)
+      setLastBlockNumber(blockNumber)
+    } finally {
+      setIsLoading(false)
+    }
   }
 
-  useEffect(() => {
+  useMemo(() => {
```

Comment:
Also shouldn't `useEffect` behave the same as `useMemo` here. Why the change ?

Reply:
Good catch. I was playing around with what worked best. I've switched back to `useEffect` as they behave the same and to also handle cleanup procedures.